### ISSUE TITLE
feat: discaurage changing default patch selection

### ISSUE THREAD

**Issue by oSumAtrIX** [revanced/revanced-manager-compose#57]
## Issue

The default selection of patches is selected by hand. The selection is made carefully to satisfy the most significant group of interest. People have been changing the selection without knowing about the consequences, such as:

- Missing a feature added by a patch
- Missing "required" patches
- Breaking the app due to including patches that are not explicitly designed for the user's selected app

## Solution

Because ReVanced Manager is agnostic to any patch and app, it can not determine a combination out of a recommended set of apps and patches. Instead ReVanced Manager can implement a warning as depicted in the following scenario:

1. The user selects an app
2. The user visits the patch selection screen - The default patches are selected
3. The user tries to change that selection
4. ReVanced Manager confronts the user with a warning. The warning can mention the pitfalls of changing the above selection or discourage the user from changing the selection.
5. The user agrees with keeping the current selection
(6. ReVanced Manager automatically redirects back from the patch selection screen to the "main" screen while confirming the default selection of patches for the user automatically.)

**Comment by Axelen123** [comment]
How will this work with MicroG when doing root install?
Having to deselect it even when manager tells you not to is not good UX

**Comment by oSumAtrIX** [comment]
If you deselect MicroG, you are gonna get a warning. Deselecting MicroG means you know what you are doing. Thus, the warning is dismissed by the user.

**Comment by Ushie** [comment]
This is not ideal, I want to remove a patch that is included by default SIMPLY because its a cosmetic patch, there are a lot of patches that are included by default for no other purpose other than cosmetic changs

Vanced MicroG comes at a bigger importance than lets say Video ads, or Hide info cards for example

And also, imagine this flow:

1. open patches selection
2. unselect lets say Video ads (because I am a premium sub for example)
3. acknowledge the warning and continue anyways
4. untick Vanced MicroG Support, which I accomplish without being warned as I had already acknowledged the warning for deselecting a patch at much lower importance than something that is **required** for primary the installation method

**Comment by Axelen123** [comment]
> If you deselect MicroG, you are gonna get a warning. Deselecting MicroG means you know what you are doing. Thus, the warning is dismissed by the user. UX is working perfectly fine in your example scenario.

No, since the warning becomes contradictory if you are doing a root install

**Comment by Ushie** [comment]
The problem resides in the fact that Vanced MicroG Support is considered on the same level as other patches, what about some acknowledgement on the side of ReVanced Patcher for when a patch is NECESSARY for installation?

**Comment by oSumAtrIX** [comment]
> > If you deselect MicroG, you are gonna get a warning. Deselecting MicroG means you know what you are doing. Thus, the warning is dismissed by the user. UX is working perfectly fine in your example scenario.
> 
> No, since the warning becomes contradictory if you are doing a root install. You **must** edit the selection to properly patch.

No, the warning is not contradictory. The warning remains correct. 

It says  "If you don't know what you are doing don't do it". Your scenario infers that the user knows what they are doing by being rooted and explicitly unticking Vanced MicroG patch for that reason (so they know what they are doing). The warning remains correct for that reason.

**Comment by Ushie** [comment]
> Vanced MicroG is not a cosmetic patch

I was putting in my perspective as a user, unticking a default patch that is merely cosmetic

**Comment by Axelen123** [comment]
Your statement and the warning are incorrect in the context of a root install. You are not told or forced to deselect it, so how will people figure that out?

**Comment by oSumAtrIX** [comment]
@Ushie I have updated my comment with additional information

**Comment by oSumAtrIX** [comment]
> Your statement and the warning are incorrect in the context of a root install. You are not told or forced to deselect it, so how will people figure that out?

No, here is a warning text formulated as originally explained:

"Changing the current selection of patches may be dangerous. If you don't know what you are doing, do not change the selection"

You claim, that the warning is incorrect in the context of a root installation.
Though I have explained in my comment that it is not incorrect. Repeating:

The user knows what they are doing in the context of a root installation. If they unselect Vanced MicroG for the reason of being rooted, they know what they are doing. The warning as described in my initial issue, says "If you don't know what you are doing, don't proceed".

**Comment by Axelen123** [comment]
And how will the user know to deselect that patch in the first place?

**Comment by oSumAtrIX** [comment]
@Axelen123 This is out of the scope of the current issue and offtopic. What do you mean?

**Comment by taku-nm** [comment]
Ideally the user would not have to visit the patch selection screen in the first place. If there is no previous selection to load, it should always just load the default selection.
If the previous selection was a prior default list which has now been updated to accommodate newer versions, it should automatically include the changes to the default list, as it is the new default.

Default should be selected already by *default* as the word implies.

**Comment by Axelen123** [comment]
> @Axelen123 This is out of the scope of the current issue and offtopic. What do you mean?

@oSumAtrIX It is in the scope of the current topic since you are suggesting we display a warning discouraging users from modifying the selection, even though that is a requirement for successfully doing a root install (microg).

Users may not know that they must deselect the patch when doing a root install. This must be handled in some way that does not have bad UX and gives users a chance to succeed on the first try.

**Comment by oSumAtrIX** [comment]
@Axelen123 

> It is in the scope of the current topic since you are suggesting we display a warning discouraging users from modifying the selection, even though that is a requirement for successfully doing a root install (microg).

No, you are leaving out an important aspect. The warning explicitly states that it is discouraged, **if you do not know what you are doing**.

> Users may not know that they must deselect the patch when doing a root install (worsened by the warning). This must be handled in some way that does not have bad UX and gives users a chance to succeed on the first try.

Root installation works fine with Vanced MicroG support patch. If the user does not know what to deselect, it is better to not deselect, even in the case of Vanced MicroG and even in the context of root. If they do know, the discouragement automatically stops to apply to them.

**Comment by Axelen123** [comment]
In that case, mention it in the documentation of the microg patch.

If this is the approach you are going for then should also make settings patches as well

**Comment by oSumAtrIX** [comment]
> In that case, mention it in the documentation of the microg patch.

Please elaborate further. Keep in mind to be precisely on topic for this issue. 

> If this is the approach you are going for then should also make settings patches as well for apps with a large amount of patches

I do not understand this sentence. What do patches with in app settings have to do with this issue?

**Comment by Axelen123** [comment]
> Please elaborate further. Keep in mind to be precisely on topic for this issue.
Mention that it is not required for root in the description. MicroG is relevant to this discussion.

> I do not understand this sentence. What do patches with in app settings have to do with this issue?

Settings become important for apps with revanced features you may want to disable (such as cosmetics). The selector provided an alternative to that, but its use is now discouraged

**Comment by oSumAtrIX** [comment]
If you want to exclude a patch, you can do so. The warning says if you don't know what you are doing. If someone unselects a cosmetic patch, they know what they are doing. But the same cannot be claimed for Vanced MicroG necessarily. This is why the warning says "If you know what you are doing".

**Comment by taku-nm** [comment]
It is likely not enough to rely on the hope that the user will read and/or understand the warning. The button to close the warning (but not dismiss it) should be the one attracting immediate attention. 
While the one to dismiss and therefore bypass the warning to deselect anyways, should be labelled "I know what I'm doing".
Similar to the PlayProtect warning but slightly less hidden.

**Comment by taku-nm** [comment]
>Since users have been sending screenshots of it often, this means that they do not read it, to begin with

Or that they do not understand what it means. After all it just tells you that non-root install is not possible with the current selection. How would user know, that he would need to select Vanced MicroG support - just from that warning? So obviously they go and ask. It does not really mean that they don't read at all.
The warning I proposed would be a lot less likely to show up in thousands of screenshots on discord, as it explains itself.

**Comment by oSumAtrIX** [comment]
> Or that they do not understand what it means. 

Exactly. The warning I proposed says, to leave the default patch selection if you do not know what they mean. Thus if they do not know what Vanced MicroG patch is, it warns them from unticking it. The screenshots show up because ReVanced Manager does not implement #57 and discourage users from deselecting what they don't know. I have explained this in my issue #57, please check the first paragraph.

**Comment by taku-nm** [comment]
I agree, my comment was just adding to the specifics how the warning dialogue should look and function. Just as a measure of further fool-proofing so to say.
See: 
>The button to close the warning (but not dismiss it) should be the one attracting immediate attention.
While the one to dismiss and therefore bypass the warning to deselect anyways, should be labelled "I know what I'm doing".

**Comment by oSumAtrIX** [comment]
Yes. This is a reasonable design choice that is further discussed after the issue has been acknowledged. A design can be proposed in detail, once the issue is acknowledged, the warning I proposed acts as the minimum requirement to satisfy solving #57.
